File: chrome/browser/ui/android/infobars/previews_infobar.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/ui/android/infobars/previews_infobar.h"

#include <memory>
#include <utility>

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "jni/PreviewsInfoBar_jni.h"

PreviewsInfoBar::PreviewsInfoBar(
    std::unique_ptr<PreviewsInfoBarDelegate> delegate)
    : ConfirmInfoBar(std::move(delegate)) {}

PreviewsInfoBar::~PreviewsInfoBar() {}

base::android::ScopedJavaLocalRef<jobject> PreviewsInfoBar::CreateRenderInfoBar(
    JNIEnv* env) {
  PreviewsInfoBarDelegate* delegate =
      static_cast<PreviewsInfoBarDelegate*>(GetDelegate());
  base::android::ScopedJavaLocalRef<jstring> message_text =
      base::android::ConvertUTF16ToJavaString(env, delegate->GetMessageText());
  base::android::ScopedJavaLocalRef<jstring> link_text =
      base::android::ConvertUTF16ToJavaString(env, delegate->GetLinkText());
  base::android::ScopedJavaLocalRef<jstring> timestamp_text =
      base::android::ConvertUTF16ToJavaString(
          env, delegate->GetStalePreviewTimestampText());
  return Java_PreviewsInfoBar_show(env, GetEnumeratedIconId(), message_text,
                                   link_text, timestamp_text);
}

// static
std::unique_ptr<infobars::InfoBar> PreviewsInfoBar::CreateInfoBar(
    infobars::InfoBarManager* infobar_manager,
    std::unique_ptr<PreviewsInfoBarDelegate> delegate) {
  return std::make_unique<PreviewsInfoBar>(std::move(delegate));
}

File: chrome/android/java/src/org/chromium/chrome/browser/infobar/PreviewsInfoBar.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.infobar;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.chrome.R;
import org.chromium.chrome.browser.ResourceId;

/**
 * An InfoBar that lets the user know that a Preview page has been loaded, and gives the user
 * a link to reload the original page. This InfoBar will only be shown once per page load.
 */
public class PreviewsInfoBar extends ConfirmInfoBar {
    private final String mTimestampText;

    @CalledByNative
    private static InfoBar show(
            int enumeratedIconId, String message, String linkText, String timestampText) {
        return new PreviewsInfoBar(
                ResourceId.mapToDrawableId(enumeratedIconId), message, linkText, timestampText);
    }

    private PreviewsInfoBar(
            int iconDrawbleId, String message, String linkText, String timestampText) {
        super(iconDrawbleId, R.color.infobar_icon_drawable_color, null, message, linkText, null,
                null);
        mTimestampText = timestampText;
    }

    @Override
    protected CharSequence getAccessibilityMessage(CharSequence defaultMessage) {
        return getContext().getString(R.string.previews_infobar_accessibility_title);
    }

    @Override
    public void createContent(InfoBarLayout layout) {
        super.createContent(layout);
        if (mTimestampText.isEmpty()) return;
        layout.getMessageLayout().addDescription(mTimestampText);
    }
}
